### PR TITLE
[REEF-226] A REEF.NET client for YARN

### PR DESCRIPTION
**Pull request by markusweimer**
* Refactored YarnJobSubmissionHandler to make the code reusable for the command line launcher for .NET Drivers.
  * Added YarnJobsubmissionClient to be launched by the .NET API
  * Updated HelloREEF to run on YARN as well. Just run `Org.Apache.REEF.Examples.HelloREEF.exe yarn` to run it.

JIRA: [REEF-226](https://issues.apache.org/jira/browse/REEF-226)

### CODE REVIEW

**Comment by jwang98052** [comment]
I am reviewing it.

File: lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs [hunk 1]
```diff
@@ -152,15 +153,28 @@ private string GetClientClasspath()
...
-            if (null == files || files.Length == 0)
+            if (files.Count == 0)
             {
                 Exceptions.Throw(new ClasspathException(
                     "Unable to assemble classpath. Make sure the REEF JAR is in the current working directory."), Logger);
             }
-            var classPath = string.Join(";", files);
-            return classPath;
+
+            var classpathEntries = new List<string>(_additionalClasspathEntries).Concat(files);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs]: Any synchronization issue while you are copying _additionalClasspathEntries to a new list and someone is calling AddToClassPath()?

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnCommandLineEnvironment.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        /// <summary>
+        /// Returns the full Path to the `yarn.cmd` file.
+        /// </summary>
+        /// <returns>The full Path to the `yarn.cmd` file.</returns>
+        internal string GetYarnCommandPath()
+        {
+            var result = Path.Combine(GetHadoopHomePath(), "bin", "yarn.cmd");
+            if (!File.Exists(result))
+            {
+                throw new FileNotFoundException("Couldn't find yarn.cmd", result);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnCommandLineEnvironment.cs]: Use  Exceptions.Throw()

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnCommandLineEnvironment.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            if (string.IsNullOrWhiteSpace(path))
+            {
+                throw new FileNotFoundException("HADOOP_HOME isn't set.");
+            }
+
+            var fullPath = Path.GetFullPath(path);
+
+            if (!Directory.Exists(fullPath))
+            {
+                throw new FileNotFoundException("HADOOP_HOME points to [" + fullPath + "] which doesn't exist.");
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnCommandLineEnvironment.cs]: Use  Exceptions.Throw()

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnCommandLineEnvironment.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        /// <summary>
+        /// Returns the full Path to HADOOP_HOME
+        /// </summary>
+        /// <returns>The full Path to HADOOP_HOME</returns>
+        internal string GetHadoopHomePath()
+        {
+            var path = Environment.GetEnvironmentVariable("HADOOP_HOME");
+            if (string.IsNullOrWhiteSpace(path))
+            {
+                throw new FileNotFoundException("HADOOP_HOME isn't set.");
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnCommandLineEnvironment.cs]: Use  Exceptions.Throw()

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnCommandLineEnvironment.cs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                    {
+                        output.Append(e.Data);
+                    }
+                };
+                process.BeginOutputReadLine();
+                process.WaitForExit();
+            }
+            else
+            {
+                throw new Exception("YARN process didn't start.");
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnCommandLineEnvironment.cs]: Use  Exceptions.Throw()

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEF.cs [hunk 6]
```diff
@@ -74,6 +76,8 @@ private static IConfiguration GetRuntimeConfiguration(string name)
                     return LocalRuntimeClientConfiguration.ConfigurationModule
                         .Set(LocalRuntimeClientConfiguration.NumberOfEvaluators, "2")
                         .Build();
+                case YARN:
+                    return YARNClientConfiguration.ConfigurationModule.Build();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEF.cs]: why not to specify NumberOfEvaluators when running on Yarn?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  private final Path path;
+
+  /**
+   * This constructor is only called by JobUploader.
+   *
+   * @param fileSystem
+   * @param path
+   * @throws IOException when the given path can't be created on the given FileSystem
+   */
+  JobFolder(final FileSystem fileSystem, final Path path) throws IOException {
```
Comment (jwang98052) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java]: Why not make it injectable?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+      LOG.log(Level.WARNING,
+          "Requested {0}MB of memory for the driver. " +
+              "The max on this YARN installation is {1}. " +
+              "Using {1} as the memory for the driver.",
+          new Object[]{requestedMemory, maxMemory});
+      amMemory = maxMemory;
+    }
+    return amMemory;
+  }
+
```
Comment (jwang98052) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: remove additional line

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    this.applicationSubmissionContext.setAMContainerSpec(YarnTypes.getContainerLaunchContext(launchCommand, this.resources));
+
+    LOG.log(Level.INFO, "Submitting REEF Application to YARN. ID: {0}", this.applicationId);
+
+    if (LOG.isLoggable(Level.FINEST)) {
+      LOG.log(Level.FINEST, "REEF app command: {0}", StringUtils.join(launchCommand, ' '));
+    }
+
+
```
Comment (jwang98052) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: remove additional line

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+          "Hadoop version is {0} or after with KeepContainersAcrossApplicationAttempts supported, will set it to true.",
+          minVersionKeepContainerOptionAvailable);
+
+      applicationSubmissionContext.setKeepContainersAcrossApplicationAttempts(true);
+    }
+
+    this.yarnClient.submitApplication(applicationSubmissionContext);
+  }
+
+
```
Comment (jwang98052) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: remove additional line

File: lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs (hunk 1 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs]: The class isn't threadsafe. Is there an annotation I can use to indicate that. On the java side, we use `@Threadsafe` to annotate classes that are.

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEF.cs (hunk 6 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEF.cs]: That parameter has no correspondent in YARN. The number of Evaluators set for the local runtime defines the size of the cluster simulated by it, not the number of Evaluators the job uses.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java (hunk 7 above)
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java]: An instance of this is instantiated only after `Path` is known. Hence, it cannot be injected.

**Comment by markusweimer** [comment]
@jwang98052 I merged in current master and addressed your comments.

File: lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/Common/JavaClientLauncher.cs]: C# doesn't have it. Usually we use lock.

**Pull request closed without merging**